proposals/p2511.md

```diff
@@ -0,0 +1,190 @@
+-   [Practical safety and testing mechanisms](/docs/project/goals.md#practical-safety-and-testing-mechanisms)
+    -   Replaces correctness warning on `if (a = b)` with a language rule.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   Providing largely the same set of symbols makes migration easier.
+    -   Some cost may be imposed by forcing a translation from assignment in
+        subexpressions to assignment as separate statements.
+
+## Alternatives considered
+
+### Allow assignment as a subexpression
+
+We could allow some or all forms of assignment as subexpressions, either with
+the same syntax or with some other syntax. For now, we are choosing to not
+pursue these options in order to gather data around how important they are. If
+we wish to allow assignment as a subexpression, we should consider adopting
+something like Python's "walrus operator" `variable := value` instead of using
+an `=` symbol.
+
+If we allow a three-term form of `for` statements, `for (init; cond; incr)`, we
+could allow assignment in the `incr` term. However, we currently do not support
+this syntax.
+
+### Allow chained assignment
+
+As a special case of assignment as a subexpression, we could allow chained
+assignment:
+
+```
+a = b = c = 0;
+```
+
+We could restrict this to only apply to simple assignment, not cases like
+`a += b -= c *= 2;`
+
+In leads issue [#451](https://github.com/carbon-language/carbon-lang/issues/451)
+it was decided that we would initially not support chained assignment.
+
+### Do not provide increment and decrement
+
+We could remove `++var;` and `--var;` in favor of `var += 1;` and `var -= 1;`.
+However, developers coming from C-family languages will expect these operators
+to exist, and they may more directly convey the intended semantics of counting
+than a `+=` operation would.
+
+### Define `$=` in terms of `$`
+
+Instead of defining `$` in terms of `$=`, we could define `$=` in terms of `$`
+and `=`. We choose not to do this because our experience from C++ is that `$=`
+can frequently be implemented more efficiently than `$`, by operating in-place
+and reusing allocated storage from the left-hand operand.
+
+However, the details of parameter passing and lifetimes in the Carbon object
+model have not been fully developed at this point, and it may be the case that
+implementing `$=` in terms of `$` and `=` doesn't impose a penalty. If that is
+the case, then this decision should be revisited.
+
+### Do not allow overloading the behavior of `=`
+
+We could define that an `=` expression always carries out these steps:
+
+-   Initialize a value of the left-hand operand's type from the right-hand
+    operand.
+-   Destroy the left-hand operand.
+-   Move the value created earlier into the left-hand operand.
+
+However, this removes some flexibility and could harm performance, for example
```

Comment:
Maybe worth mentioning that it'd also much more sharply deviate from C++ where types are currently designed with this flexibility in mind?

Reply:
Done.